sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,808 @@
+        do_something_else(idx, temp);
+      });
+    });
+
+    // memory is available to be used by another allocation
+    syclex::submit_with_event(q, [&](handler &cgh) {
+      cgh.depends_on(e5);
+      syclex::async_free(cgh, temp);
+    });
+  }
+
+  // memory pool releases memory back to the memory provider
+  q.wait();
+}
+----
+
+Another example of memory pool usage is described below. In this example rather
+than creating an explicit memory pool the default memory pool is being used
+instead. There is also additional queue synchronization between the commands
+enqueued which ordinarily could cause the memory to be released back to the
+memory provider, however, the reserve threshold for the memory pool is extended
+so the runtime will attempt to maintain that memory reservation and therefore
+still provide the benefit of re-allocating memory from the memory pool.
+
+[source,c++]
+----
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main(int argc, char *argv[])
+{
+  queue q(property::queue::in_order);
+
+  syclex::memory_pool memPool
+    = q.get_context().ext_oneapi_get_default_memory_pool(
+        q.get_device(), usm::alloc::device);
+
+  memPool.increase_threshold_to(1024);
+  
+  {
+    // memory pool allocates memory from the memory provider
+    void *temp = syclex::async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for first kernel
+    syclex::parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    syclex::async_free(q, temp);
+  }
+
+  // memory pool does not release memory back to the memory provider as it is
+  // still within the specified reserve threshold
+  q.wait();
+
+  {
+    // memory pool re-uses previously allocated memory
+    void *temp = syclex::async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for second kernel
+    syclex::parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something_else(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    syclex::async_free(q, temp);
+  }
+
+  // again memory pool does not release memory back to the memory provider
+  q.wait();
+}
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ASYNC_MEMORY_ALLOC` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
```

Comment:
Should this extension be versioned if it is growing over time?  Or are experimental specs not versioned as they're expected to change?